File: consumer.js
#!/usr/bin/env node

console.log("Javascript-Worker (consumer) Brute-Force cracking pwd hashes...");

const CONN_URL = 'lxmkvcva:lcUgTmzJKnpyiVwMKTOcjmXfMzWnfQfs@sparrow.rmq.cloudamqp.com/lxmkvcva';
var crackqueue = 'cracking';
var response = '';
var crypto = require("crypto-js");

//Erstellt alle Möglichten Kombinationen
function allPossibleCombinations(input, length, curstr) {
    if(curstr.length == length) return [ curstr ];
    var ret = [];
    for(var i = 0; i < input.length; i++) {
        ret.push.apply(ret, allPossibleCombinations(input, length, curstr + input[i]));
    }
    return ret;
}

var all = ['0','1','2','3','4','5','6','7','8','9','a','b','c','d','e','f','g','h','i','j','k','l','m','n','o','p','q','r','s','t','u','v','w','x','y','z']

var amqp = require('amqplib/callback_api');
//Verbindung zu RabbitMQ Herstellen
amqp.connect('amqp://' + CONN_URL, function(error0, connection) {
  if (error0) {
    throw error0;
  }
  //Channel Erstellen
  connection.createChannel(function(error1, channel) {
    if (error1) {
      throw error1;
    }
    var queue = crackqueue;

    //Queue dem Channel zuweisen
    channel.assertQueue(queue, {
      durable: false
    });
    //Auf Message warten
    channel.consume(queue, function reply(msg) {
      var hash = msg.content.toString();
      console.log(" [.] hash(%d)", hash);
    
      //Alle Möglichkeiten erstelllen (1,2 und 3 Stellig)
      var combinationsWithOne = allPossibleCombinations(all,1,'')
      var combinationsWithTwo = allPossibleCombinations(all,2,'')
      var combinationsWithThree = allPossibleCombinations(all,3,'')
      
      var savedCombinations = []
      //in ein Array zusammenfügen
      savedCombinations.push(...combinationsWithOne, ...combinationsWithTwo, ...combinationsWithThree)
  
      //Eine Schleife über das Array laufen lassen und schauen, ob Hash der an der Stelle von i ist
	for(var i = 0; i<=savedCombinations.length; i++){
		var current_hash = crypto.MD5(savedCombinations[i]);
		console.log(savedCombinations[i] + " is the current hash: " + current_hash + " == " + hash + " hash ??");
    //Bei Übereinstimmung, den Wert in response speichern
    if (current_hash == hash){
			console.log("(js) Found password: " + savedCombinations[i]);
      response = (""+savedCombinations[i]);
      //über den Channel zurückschicken
			channel.sendToQueue(msg.properties.replyTo,
				Buffer.from(response), {
					correlationId: msg.properties.correlationId
				});
			channel.ack(msg);
			break;
		}
	}
    });
  });
});

File: consumer.py
#!/usr/bin/python3

import pika
import hashlib
import os

print("Python-Worker (consumer) Brute-Force cracking pwd hashes...")

url='amqp://lxmkvcva:lcUgTmzJKnpyiVwMKTOcjmXfMzWnfQfs@sparrow.rmq.cloudamqp.com/lxmkvcva'
port=1883
crackqueue = 'cracking'
response = ''

def baseN(num, b=36, numerals="0123456789abcdefghijklmnopqrstuvwxyz"):
    return ((num == 0) and numerals[0]) or (baseN(num // b, b, numerals).lstrip(numerals[0]) + numerals[num % b])


def crack_callback(ch, method, properties, body):
        print("TODO please crack pwd-hash {body}")
        for i in range(36**3):
                current_hash = hashlib.md5(str(baseN(i).zfill(1)).encode()).hexdigest()
                print(f"{baseN(i)} is the current hash {current_hash} == {body} hash ??")
               
                if current_hash == body.decode('utf-8'):
                        print(f"(py) Found Password: {baseN(i)}")
                        response = baseN(i)
                        ch.basic_publish(
                            exchange='', 
                            routing_key=properties.reply_to, 
                            properties=pika.BasicProperties(correlation_id = properties.correlation_id), 
                            body=str(response))
                        ch.basic_ack(delivery_tag=method.delivery_tag)                     
                        break



conn = pika.BlockingConnection(pika.URLParameters(url))
channel = conn.channel()
channel.queue_declare(queue=crackqueue)

channel.basic_consume(queue = crackqueue, on_message_callback=crack_callback)
channel.start_consuming()
conn.close()
print("Done")
